### PR TITLE
feat(dex): add support for solidly exchange

### LINKED ISSUE #5820: [Bounty] Integrate Solidly, a DEX on Fantom into Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks 
- Bounty Size
  - [x] 1000 USDC or $MASK equivalent 

## Part 1: Problem Description
**Background**  Solidly is an AMM based on Uniswap v2. It allows swapping between two tokens in a liquidity pool in a fully decentralized way. The first initial twist is that Solidly, unlike Uniswap v2, also supports low slippage trades between pegged assets like stablecoins. Users who provide their liquidity in a whitelisted Solidly pool receive liquidity mining rewards in the form of SOLID, the governance and revenue token for the Solidly protocol.

Solution Create a plugin that integrates Solidly into Mask. The UI should be identical to other swaps.

image

image

_Solidly Swap Page_ 

Screen Shot 2022-03-07 at 4 54 02 PM

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://github.com/solidlyexchange/solidly
- https://solidly.exchange/
- https://github.com/solidlyexchange/solidly

## Part 3: Contact and Support
@yisiliu @tzhan28 @LindaMask for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q,
Solidly Discord: https://discord.com/invite/spartacadabra

Please speak up in the Mask discord's bounty section or just discord dm @yisiliu or @taylor zhang or @lin_dao #5820  so that we can find you on Discord. For faster communication reach out to lin_DAO on telegram and she will create a dedicated group with the Mask team to start the bounty process

Submission Requirements
Please include demo to show the integration is working.

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by brymut**
## Description

Integrate Solidly Exchange for Swapping in Mask Plugin on Fantom.

Closes #5820 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

**Comment by UncleBill** [comment]
@brymut 
It's very different to official website.

image

image

File: packages/mask/src/plugins/Trader/helpers/pair.ts [hunk 1]
```diff
@@ -22,7 +28,9 @@ export function getPairAddress(factoryAddress: string, initCodeHash: string, tok
                 ...cache?.[token0Addr],
                 [token1Addr]: getCreate2Address(
                     factoryAddress,
-                    keccak256(['bytes'], [pack(['address', 'address'], [token0Addr, token1Addr])]),
+                    stable
+                        ? keccak256(['bytes'], [pack(['address', 'address', 'bool'], [token0Addr, token1Addr, stable])])
+                        : keccak256(['bytes'], [pack(['address', 'address'], [token0Addr, token1Addr])]),
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/helpers/pair.ts:33]: Could you explain why adding a `stable` parameter?

File: packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts [hunk 2]
```diff
@@ -31,7 +31,7 @@ export function usePairs(tradeProvider: TradeProvider, tokenPairs: readonly Toke
         if (!FACTORY_CONTRACT_ADDRESS || !INIT_CODE_HASH) return EMPTY_LIST
         return tokenPairs.map(([tokenA, tokenB]) =>
             tokenA && tokenB && !tokenA.equals(tokenB)
-                ? getPairAddress(FACTORY_CONTRACT_ADDRESS, INIT_CODE_HASH, tokenA, tokenB)
+                ? getPairAddress(FACTORY_CONTRACT_ADDRESS, INIT_CODE_HASH, tokenA, tokenB, stable)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts]: A property passing down is unnecessary. Prefer to use `context.STABLE`. Yes, treat the `STABLE` as a common configuration of trade providers.

File: packages/mask/src/plugins/Trader/helpers/pair.ts (hunk 1 above)
Comment (guanbinrui) [reply packages/mask/src/plugins/Trader/helpers/pair.ts:33]: Hi @brymut, could you please paste a source code link here?

Comment (brymut) [reply packages/mask/src/plugins/Trader/helpers/pair.ts:33]: Hi, please check the code for the deployed contract, `stable` bool required to obtain pair addresses in `pairFor` function https://ftmscan.com/address/0xa38cd27185a464914D3046f0AB9d43356B34829D#code

File: packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts (hunk 2 above)
Comment (brymut) [reply packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts]: Changed

**Comment by guanbinrui** [comment]
Hi @brymut Plz resolve those conflicts.

**Comment by guanbinrui** [comment]
Hi @brymut, any updates?

**Pull request closed without merging**